common/src/main/java/com/proximyst/ban/commands/HistoryCommand.java

```diff
@@ -63,17 +65,31 @@ public void register(final @NonNull CommandManager<@NonNull IBanAudience> comman
   private void execute(final @NonNull CommandContext<IBanAudience> ctx) {
     final BanUser target = ctx.get("target");
 
-    this.messageService.sendFormattedMessage(ctx.getSender(), Identity.nil(), MessageKey.COMMANDS_FEEDBACK_HISTORY,
-        "targetName", target.getUsername(),
-        "targetUuid", target.getUuid())
-        .exceptionally(this.banExceptionalFutureLogger.cast());
+    this.messageService.commandsFeedbackHistory(target)
+        .send(ctx.getSender());
 
     this.punishmentService.getPunishments(target.getUuid())
-        .thenCompose(punishments -> this.messageService.formatHistory(punishments, target))
-        .thenAccept(messages -> {
-          for (final Component message : messages) {
-            ctx.getSender().sendMessage(Identity.nil(), message);
+        .thenCompose(punishments -> {
+          final CompletableFuture<?>[] futures = new CompletableFuture<?>[punishments.size()];
+          int index = 0;
+          for (final Punishment punishment : punishments) {
+            futures[index++] = this.messageService.commandsHistoryEntry(punishment)
+                .component();
           }
+
+          return CompletableFuture.allOf(futures)
+              .thenApply($ -> Arrays.stream(futures)
+                  .map(f -> (Component) f.join())
+                  .toArray(Component[]::new));
+        })
+        .thenAccept(messages -> {
+          this.messageService.commandsHistoryHeader(target, messages.length)
+              .send(ctx.getSender())
+              .thenRun(() -> {
+                for (final Component message : messages) {
+                  ctx.getSender().sendMessage(Identity.nil(), message);
+                }
+              });
```

Comment:
I would personally extract the thenCompose and thenAccept lambdas to their own methods, this makes it easier to see what is happening at a glance without having to read a ton of code.